src/main/scala/nl/knaw/dans/api/sword2/DepositHandler.scala

```diff
@@ -160,14 +166,61 @@ object DepositHandler {
     }
   }
 
-  private def checkBagValidity(bagitDir: File)(implicit id: String): Try[Unit] = {
+  def resolveFetchItems(bagitDir: File)(implicit id: String): Try[Unit] = Try {
+    log.debug(s"[$id] Checking if bag contains fetch.txt")
+    getFetchTxt(bagitDir).foreach(fetchText =>
+      { checkFetchTxtUrls(fetchText)
+        resolveFetchTxtFiles(bagitDir, fetchText)
+      }
+    )
+  }
+
+  private def getFetchTxt(bagitDir: File)(implicit id: String): Option[FetchTxt] = Option {
+    getBagFromDir(bagitDir).getFetchTxt
+  }
+
+  private def checkFetchTxtUrls(fetchText: FetchTxt)(implicit id: String): Unit = {
+    log.debug(s"[$id] Checking validity of urls in fetch.txt")
+    fetchText.asScala.foreach(item =>
+      checkUrlValidity(item.getUrl))
+  }
+
+  private def checkUrlValidity(url: String)(implicit id: String): Unit = {
+    // check if the url is syntactically correct
+    try {
+      new URL(url)
+    } catch {
+      case e: Throwable => throw new InvalidDepositException(id, s"Invalid url syntax in Fetch Items ($url)")
+    }
+    // check if the url complies with the allowed url-structure
+    val urlPattern = Pattern.compile(SwordProps("url-pattern"))
+    if (!urlPattern.matcher(url).matches)
+      throw new InvalidDepositException(id, s"Not allowed url in Fetch Items ($url)")
+  }
+
+  private def resolveFetchTxtFiles(bagitDir: File, fetchText: FetchTxt)(implicit id: String): Unit = {
+    log.debug(s"[$id] Resolving files in fetch.txt")
+    fetchText.asScala.foreach(item =>
+      Using.urlInputStream(new URL(item.getUrl)).foreach(src =>
+        Files.copy(src, Paths.get(bagitDir.getAbsolutePath, item.getFilename))))
+  }
+
+  def checkBagValidity(bagitDir: File)(implicit id: String): Try[Unit] = {
     log.debug(s"[$id] Verifying bag validity")
-    val bag = bagFactory.createBag(bagitDir, BagFactory.Version.V0_97, BagFactory.LoadOption.BY_MANIFESTS)
+    val bag = getBagFromDir(bagitDir)
     val validationResult: SimpleResult = bag.verifyValid
     if (validationResult.isSuccess) Success(Unit)
     else Failure(new InvalidDepositException(id, validationResult.messagesToString))
   }
 
+  private def getBagFromDir(dir: File)(implicit id: String): Bag = {
+    try {
+      bagFactory.createBag(dir, BagFactory.Version.V0_97, BagFactory.LoadOption.BY_MANIFESTS)
+    } catch {
+      case e: Throwable => throw new FailedDepositException(id, "Failed to create a bag", e)
```

Comment:
I actually had first `scala.util.Try`.  But there is no advantage of that here, because these exceptions cause the whole process to stop.